propolis/src/dispatch/mod.rs

```diff
@@ -361,6 +387,10 @@ impl DispCtx {
     ) -> DispCtx {
         DispCtx { mctx, event: EventCtx::new(edisp), ctrl, di: inner }
     }
+
+    /// Returns true if the function holding this [`DispCtx`] object
+    /// should yield control back to the dispatcher.
+    // TODO: Why "&mut self", not just "&self"?
```

Comment:
The lifetime requirement means there would be no "dangling borrows" of resources from this DispCtx at the time of yield.

Reply:
I mean to resources exposed by the `DispCtx`.  I think it's plausible that things like the `memctx` other resource maps are updated with less lock-heavy implementation in the future.  The dispatcher yield point is a natural fit for how/when to update them.  The `&mut` prevents the worker thread from holding onto references which might become stale.

This example shows a case where those references held across the `check_yield()` aren't allowed:
```rust
    fn process_loop(&self, ctx: &mut DispCtx) {
        let mut reqs = self.reqs.lock().unwrap();
        loop {
            let mem = ctx.mctx.memctx();

            let some_data: u32 = mem.read(GuestAddr(0x1000)).unwrap();
            if ctx.check_yield() {
                break;
            }
            let more_data: u32 = mem.read(GuestAddr(0x1000)).unwrap();
...
```
The compiler balks about the mutable borrow at `check_yield`.

Switched up a little bit, the story is different, since `memctx` wasn't held across the yield point:
```rust
    fn process_loop(&self, ctx: &mut DispCtx) {
        let mut reqs = self.reqs.lock().unwrap();
        loop {
            let some_data: u32 = ctx.mctx.memctx().read(GuestAddr(0x1000)).unwrap();
            if ctx.check_yield() {
                break;
            }
            let more_data: u32 = ctx.mctx.memctx().read(GuestAddr(0x1000)).unwrap();
...
```

Reply:
Interesting - I think the intent here is a little more clear to me: if the processing loop decides to yield, it should not be holding context from before the yield point, and forcing access to a `&mut self` method in `DispCtx` forces all other lifetimes referencing `DispCtx` to end (is that right?).

I think this goal makes sense, although I'm struggling a little with this mechanism for enforcement.

1) The `check_yield` method isn't *really* what you're trying to use as the "end of `&DispCtx` lifetimes" - the *yield* point would be where you want the lifetime to end.
2) A caller could choose to yield unconditionally, in which case the `&mut self` enforcement would never occur.
3) Even in the cases mentioned above, the yield happens to be implemented as "return from the processing function" (basically the `break` statement). In those cases, it isn't possible to "access `some_data`, yield, and then access `more_data`" - if the yield happens, we've already left the loop. I think this is similar to point (1) though, of focusing on "what does a yield look like" rather than "what does a check-if-we-should-yield look like".

Reply:
Spoke offline (well, online, you know how it is) and resolved this - I was totally wrong, `break` isn't the yield, `check_yield` is the yield, so it makes sense that it needs to act on `&mut self`.